src/meta/expr.lean

```diff
@@ -372,6 +372,35 @@ e.replace $ λ e d,
   | _ := none
   end
 
+/-- Implementation of `expr.mreplace`. -/
+meta def mreplace_aux {m : Type* → Type*} [monad m] [alternative m] (R : expr → nat → m expr) :
```

Comment:
I think it would be better to have the function take `R : expr -> nat -> m (option expr)` so that the monad doesn't have to be alternative (and failure, if it exists, cancels the whole traversal). You can recover this version by composing with `try_core`.

Reply:
I've made this change, but I now have:
```
/-- FIXME err, what is this called, and is there just a punctuation incantation I can use? -/
def foo {α : Type*}{m : Type* → Type*} [monad m] (x : m (option α)) (y : m α) : m α :=
do o ← x,
  match o with
  | none := y
  | (some a) := pure a
  end
```